Add section on span status for databases (#1560)
open-telemetry/semantic-conventions: docs/database/database-spans.md

## Status

[Span Status Code][SpanStatus] MUST be left unset if the operation has ended without any errors.

Instrumentation SHOULD consider the operation as failed if any of the following is true:

- the `db.response.status_code` value indicates an error

  > [!NOTE]
  > The classification of status code as an error depends on the context.
  > For example, a SQL STATE `02000` (`no_data`) indicates an error when the application
  > expected the data to be available. However, it is not an error when the
  > application is simply checking whether the data exists.
  >
  > Instrumentations that have additional context about a specific operation MAY use
  > this context to set the span status more precisely.
  > Instrumentations that don't have any additional context MUST follow the
  > guidelines in this section.

- an exception is thrown by the instrumented method call
- the instrumented method returns an error in another way

When the operation ends with an error, instrumentation:

- SHOULD set the span status code to `Error`
- SHOULD set the `error.type` attribute

> - SHOULD set the span status description when it has additional information

Reviewer: should this have any requirements about not having any confidential data? Similar to how we have on metrics errors where we don't add if it shows actual values of queries, etc

Author: I don't see how database instrumentation can determine if there is any sensitive or confidential data in the arbitrary exception message.  Someone can put sensitive data into database name (bluesky has DB per user and they might call them accordingly), but how can we provide basic observability in the common case if we assume the worst? 

Happy to consider any specific cases when we can do something on the instrumentation side.

Reviewer: Sorry if I wasn't clear, I didn't mean that is the job of the instrumentation to determine this, I meant more as a recommendation, that this sentence could include something on the lines of 

```
SHOULD set the span status description when it has additional information about the error that aligns with [Span Status Description][SpanStatus] and does not contain sensitive data.
```

  about the error that aligns with [Span Status Description][SpanStatus].

  It's NOT RECOMMENDED to duplicate `db.response.status_code` or `error.type`
  in span status description.

  When the operation fails with an exception, the span status description SHOULD be set to
  the exception message.

### Recording exception events

**Status**: [Experimental][DocumentStatus]

When the operation fails with an exception, instrumentation SHOULD record

> an [exception event](../exceptions/exceptions-spans.md) by default if and only if the

Reviewer 2: ```suggestion
an [exception event](../exceptions/exceptions-spans.md) by default if, and only if, the
```

span being recorded is a local root span (does not have a local parent).

> [!NOTE]

> > Exception stack traces could be very long and are expensive to capture and store.

Reviewer 2: ```suggestion
>
> Exception stack traces could be very long and are expensive to capture and store.
```
I think you need an empty line here so the icon shows up.

> Exceptions which are not handled by instrumented libraries are likely to be handled
> and logged by the caller.
> Exceptions that are not handled will be recorded by the outermost (local root)

> > instrumentation such as HTTP or gRPC server.

Reviewer 3: let's ask for feedback on this point from the Specification SIG before we merge

Reviewer 4: I think this is a directional shift in logging exceptions.

I see two things here:

1. We still can't determine "layers" of instrumentation, i.e. we *can't* know if someone else will log the exception or not.
2. This violates the previously recorded wisdom of how to handle exceptions and stack traces.

My opinion:

This is the right direction to go.  We *should* assume that there is "outer" instrumentation that logs/catches exceptions, particularly for database "client" spans.  I.e. we may be able to blanketly state that "client" spans should never log unhandled exceptions and that we encourage "outer" instrumentation that will (e.g. server spans).

Reviewer 3: I added to next week's spec SIG agenda

Reviewer 3: Feedback from the spec SIG was that this sounds like a good default behavior, but we should make it configurable.

Some options for making it configurable:

(A) handle this all in the SDK:
* new configuration added to [TracerConfig](https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/trace/sdk.md#tracerconfig)
* [RecordException](https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/trace/api.md#record-exception) behavior could depend on this new configuration in TracerConfig

but I don't think we can change the default behavior of the SDK.

(B) Another option could be to introduce "trace advice" (similar to metric advice) that would allow a given instrumentation the ability to opt-in to this new behavior.

(C) Another option could be:
* new configuration added to [TracerConfig](https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/trace/sdk.md#tracerconfig)
* instead of changing the (existing) behavior of RecordException, ask instrumentations to check this configuration value before calling RecordException

(D) Another option could be:
* Introduce a standard "database" instrumentation configuration property (similar to the standard "http" configuration properties that we already have: https://github.com/open-telemetry/opentelemetry-configuration/blob/e1f89ef021f3c8e000d05e6b6e53bc4a2372fe52/examples/kitchen-sink.yaml#L396-L400)
* ask instrumentations to check this configuration value before calling RecordException
* we may also want to be able to override this at individual instrumentation level somehow
 
(E) Or even:
* Introduce a global instrumentation configuration property that covers all instrumentations
* ask instrumentations to check this configuration value before calling RecordException
* we may also want to be able to override this at individual instrumentation level somehow

Probably we should spin this out into a separate issue.

cc @jack-berg

Author: Thanks for the notes!

Since we don't usually specify how instrumentations should configure things, I'd prefer to tackle it separately. 
It's definitely a problem that we should fix in the https://github.com/open-telemetry/semantic-conventions/issues/1536 and https://github.com/open-telemetry/semantic-conventions/issues/1450

If a corresponding `{target}` value is not available for a specific operation, the instrumentation SHOULD omit the `{target}`.
For example, for an operation describing SQL query on an anonymous table like `SELECT * FROM (SELECT * FROM table) t`, span name should be `SELECT`.

## Status

> [Span Status Code][SpanStatus] MUST be left unset if operation has ended without any errors.

Reviewer 3: ```suggestion
[Span Status][SpanStatus] MUST be left unset if the operation has ended without any errors.
```

> Instrumentation SHOULD consider operation as failed if:

Reviewer 3: ```suggestion
Instrumentation SHOULD consider the operation as failed if any of the following is true:
```

- the `db.response.status_code` value indicates an error

  > [!NOTE]
  >
  > The classification of status code as an error depends on the context.

>   > For example, a SQL STATE `no_data` status code indicates an error when the application

Author: ```suggestion
  > For example, a SQL STATE `02000` (`no_data`) status code indicates an error when the application
```

>   > For example, a SQL STATE `no_data` status code indicates an error if the application
>   > expected the data to be available. However, it is not an error when the

Reviewer 3: (just to match the `when` in the next sentence)
```suggestion
  > For example, a SQL STATE `no_data` status code indicates an error when the application
  > expected the data to be available. However, it is not an error when the
```

  > application is simply checking whether the data exists.
  >
  > Instrumentations that have additional context about a specific operation MAY use
  > this context to set the span status more precisely.
  > Instrumentations that don't have any additional context MUST follow the
  > guidelines in this section.

> - exception is thrown by the instrumented method call

Reviewer 3: ```suggestion
- an exception is thrown by the instrumented method call
```

> - instrumented method returns error in another way

Reviewer 3: ```suggestion
- the instrumented method returns error in another way
```

> - the instrumented method returns error in another way

Reviewer 3: ```suggestion
- the instrumented method returns an error in another way
```

> When operation ends with an error, instrumentation:

Reviewer 3: ```suggestion
When the operation ends with an error, instrumentation:
```

> - SHOULD set span status code to `Error`

Reviewer 3: ```suggestion
- SHOULD set the span status code to `Error`
```

> - SHOULD set `error.type` attribute

Reviewer 3: ```suggestion
- SHOULD set the `error.type` attribute
```

> - SHOULD set span status description when it has additional information

Reviewer 3: ```suggestion
- SHOULD set the span status description when it has additional information
```

  about the error that aligns with [Span Status Description][SpanStatus] and

>   can't be inferred from `db.response.status_code` or `error.type` attributes.

Reviewer 3: I didn't follow why this part? thanks

> and can't be inferred from `db.response.status_code` or `error.type` attributes

Author: I'm trying to avoid duplicating things between `db.response.status_code` / `error.type` and span status description.

We have a similar wording in HTTP 

https://github.com/open-telemetry/semantic-conventions/blob/43f9e308ada81c5ef6851cd665d95f294b157a0f/docs/http/http-spans.md?plain=1#L112

Let me see if I can explain it more clearly

Reviewer 3: > We have a similar wording in HTTP

👍

I was thinking it might be useful for backends which display error status + description in a general way across all spans, but definitely understand the duplication concern

Reviewer 3: I'm good resolving, makes sense to stay consistent with HTTP

>   When operation fails with exception, the span status description SHOULD be set to
>   exception message.

Reviewer 3: ```suggestion
  When the operation fails with an exception, the span status description SHOULD be set to
  the exception message.
```

### Recording exception events

**Status**: [Experimental][DocumentStatus]

> When the operation fails with an exception, instrumentation MAY record

Reviewer 3: what are your thoughts?
```suggestion
When the operation fails with an exception, instrumentation SHOULD record
```

> When operation fails with exception, instrumentation MAY record
> [exception event](../exceptions/exceptions-spans.md) by default if and only if the

Reviewer 3: ```suggestion
When the operation fails with an exception, instrumentation MAY record
an [exception event](../exceptions/exceptions-spans.md) by default if and only if the
```

span being recorded is a local root span (does not have a local parent).

> [!NOTE]
> Exception stack traces could be very long and are expensive to capture and store.
> Exceptions which are not handled by instrumented libraries are likely to be caught,
> logged, and handled by the caller.

> > Exceptions that are not handled are recorded by they outermost instrumentation

Reviewer 3: ```suggestion
> Exceptions that are not handled will be recorded by the outermost (local root) instrumentation
```

> such as HTTP or gRPC server.

Instrumentations SHOULD NOT record exception events (by default) on spans that

> have local parents.

Reviewer 3: this seems covered by the "if and only if" above?
